### PR TITLE
Allow configuration name as prefix

### PR DESCRIPTION
**Pull request by hrittich**
Hi,

I am currently working on packaging Syncthing Tray for Debian, continuing Nicholas' work. Concerning #16, I have developed a set of patches that allow the configuration name feature to also be used to set a prefix instead of a suffix. Would you be willing to merge these changes into your library? It would make packaging easier for me. If you agree, I would then also send pull requests for qtutilities and Syncthin Tray.

### CODE REVIEW

**Review by Martchus — COMMENTED** [review]
Besides the point that this should be an orthogonal feature this looks like something which would be generally acceptable. Note that further changes might be required in `qtutilities` and `syncthingtray` itself.

Note that introducing such a configuration raises the problem of being able to actually find the library when building another library/application consuming it. Therefore there also needs to be  support from the side of the consuming projects, e.g. `qtutilities` has currently an additional option to specify the suffix for `c++utilities`: https://github.com/Martchus/qtutilities/blob/836801fb3967a25e6366b0dd380b205458477e94/CMakeLists.txt#L172

By the way, the next release will require the additional dependency `qtforkawesome` and the *current* release is actually still bundling icons from Font Awesome which might be problematic for Debian. So you most likely want to wait for the next release anyways.

**Comment by Martchus** [comment]
So it would make sense to:

* Make it an *addition* to the suffix
* Go ahead and checkout the other repositories (`qtutilities`, `qtforkawesome`, `syncthingtray`) and add possibly required code there and also take care that these repositories can actually find each other.
* Test this by installing and consuming the libraries/applications to/from a custom install prefix.

By the way:
* If you're going though the code you'll notice that I implemented the possibility to specify a different suffix for `qtutilities` and `qtforkawesome` because I want a different suffix for the variants built against Qt 6 but still use the regular `c++utilities`. For the prefix I would refrain from such a distinction because there's likely no use case for it, e.g. you likely want to use the same "namespace" for all my projects.
* If you decide to link against Qt 6 (instead of Qt 5) the Dolphin/Plasma plugin cannot be built and Qt 5 integration/style plugins cannot be used. So I tend to stick to Qt 5 for my "regular" packages and only offer Qt 6 variants in addition.

**Comment by hrittich** [comment]
> Besides the point that this should be an orthogonal feature this looks like something which would be generally acceptable. Note that further changes might be required in `qtutilities` and `syncthingtray` itself.

I am aware of this issue. I have already included those changes in qtutilities and Syncthing Tray. I just wanted to hear your opinion before flooding you with pull requests for each of those repositories.

> I would rather allow specifying the prefix in addition to a suffix. These features should be orthogonal.

In this case, I would suggest to add a `NAMESPACE` CMake variable to configure the prefix and to add a `PACKAGE_NAMESPACE` to the dependent projects. This would make it clear that these two settings are orthogonal.

**Comment by Martchus** [comment]
> I just wanted to hear your opinion before flooding you with pull requests for each of those repositories.

Good approach :-)

> In this case, I would suggest to add a NAMESPACE CMake variable to configure the prefix and to add a PACKAGE_NAMESPACE to the dependent projects. This would make it clear that these two settings are orthogonal.

Ok, the variable names make sense.

File: cmake/modules/BasicConfig.cmake [hunk 1]
```diff
@@ -76,7 +83,14 @@ if (${META_PROJECT_VARNAME_UPPER}_CONFIGURATION_TARGET_SUFFIX STREQUAL "none")
 elseif (${META_PROJECT_VARNAME_UPPER}_CONFIGURATION_TARGET_SUFFIX)
     set(TARGET_SUFFIX "-${${META_PROJECT_VARNAME_UPPER}_CONFIGURATION_TARGET_SUFFIX}")
 elseif (CONFIGURATION_TARGET_SUFFIX)
-    set(TARGET_SUFFIX "-${CONFIGURATION_TARGET_SUFFIX}")
+    set(TARGET_SUFFIX "-${CONFIGURATION_TARGET_SUFFIX}-")
```
Comment (Martchus) [inline cmake/modules/BasicConfig.cmake:86]: This line looks wrong. The suffix shouldn't end with a dash.

**Review by Martchus — CHANGES_REQUESTED** [review]
Looks good except for one line (see inline comment).

I assume you've simply searched for `META_CONFIG_SUFFIX` and `TARGET_SUFFIX` to identify all the places where an adjustment is needed. If that's the case everything should be covered.

File: cmake/modules/BasicConfig.cmake (hunk 1 above)
Comment (hrittich) [reply cmake/modules/BasicConfig.cmake:86]: Please ignore this push. I somehow messed up.

**Comment by hrittich** [comment]
So, this would be my suggestion for implementing the feature. What do you think?

**Review by Martchus — APPROVED** [review]
Looks good now.

**Comment by sten0** [comment]
Thank you @hrittich, much appreciated! :-)

**Pull request merged**